KAFKA-16450: Add instance methods to Exit to replace static methods

gharris1727 opened the pull request:
This adds a thread-safe alternative to the current Exit implementation, which doesn't make use of static mutable fields. This solves a number of problems:

1. Thread isolation: Two tests can inject exit procedures concurrently, without conflicting with one-another
2. The test-only methods for manipulating the state of Exit are present on the main class, and so could be called in a non-test environment.
3. When fatal procedures are not intercepted by Exit, shutdown hooks are delayed until the real JVM exit, possibly leaking resources/files across tests.
4. When fatal procedures are intercepted by Exit, those exceptions can go un-noticed, and may not cause the test to fail.
5. When shutdown hooks are intercepted by Exit, they are dropped and never executed
6. It isn't clear to code calling exit() that an exception might be thrown instead when the procedures are mocked. 

The Java and Scala Exit classes don't use any instance methods, so I was able to directly add instance methods to the existing classes. This should prevent the need for a large rename later, such as appears necessary with the other options I could think of:

* Add it as a different class SafeExit: large amount of code churn when Exit is removed, silly name
* Moving "Exit" to "UnsafeExit" now: large amount of code churn up-front, immediate merge conflicts
* Add it to a different package as "utils.temp.Exit": moderate amount of code churn when Exit is removed, harder to review

Because static and instance methods cannot share signatures, I had to perturb the names slightly. I think the new names are acceptable, and shouldn't cause any additional code churn.
* exit to exitOrThrow
* halt to haltOrThrow
* addShutdownHook to addShutdownRunnable

In the interest of not testing tests, I've omitted tests for these changes. The existing ExitTest is testing the mocking methods, which are not part of the main API now, just the MockExit API.

I imagine that this will be rolled out in stages:

1. Replace Exit.exit() calls (and similar) in Kafka with Exit.staticContext().exitOrThrow().
2. Move the Exit.staticContext() calls upward in the call graph to evaluate the staticContext() as soon as possible, and ensure that objects use consistent exit procedures
3. Replace Exit.setExitProcedure in Kafka tests with MockExit
4. Deprecate thread-unsafe Exit procedures as a signal to the community
5. Remove thread-unsafe Exit procedures, replacing Exit.staticContext() with Exit.system()

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

gharris1727 committed:
KAFKA-16420: Add instance methods to Exit to replace static methods

Signed-off-by: Greg Harris <greg.harris@aiven.io>

C0urante commented on clients/src/main/java/org/apache/kafka/common/utils/Exit.java:
```diff
@@ -119,4 +119,51 @@ public static void resetHaltProcedure() {
         shutdownHookAdder = DEFAULT_SHUTDOWN_HOOK_ADDER;
     }
+
+    public void exitOrThrow(int statusCode) {
+        exit(statusCode, null);
+    }
+
+    public abstract void exitOrThrow(int statusCode, String message);
+
+    public void haltOrThrow(int statusCode) {
+        halt(statusCode, null);
+    }
+
+    public abstract void haltOrThrow(int statusCode, String message);
+
+    public abstract void addShutdownRunnable(String name, Runnable runnable);
```
Could we have brief Javadocs on these? Beyond the obvious, it'd be nice to note when callers could expect the `OrThrow` behavior to take place.

C0urante reviewed (commented):
Thanks Greg. I like the balance this strikes between being effective and non-invasive, and the small name changes seem fine.

In order to gauge the utility of the new `MockExit` class, would it be possible to draft a commit or two that demonstrates how they might be introduced into the code base? Doesn't have to be part of this PR, can be a separate one that's based on top of the branch for this one.

gharris1727 commented:
Thanks @C0urante for the review. I put together a commit to apply this new API to the Connect module, you can see it on this branch: https://github.com/gharris1727/kafka/commits/proto-connect-exit/

I added the Exit.staticContext() method to allow partial rollout of this feature, in situations where no Exit instance is explicitly specified. Let me know what you think.

C0urante commented on clients/src/test/java/org/apache/kafka/common/utils/MockExit.java:
```diff
@@ -0,0 +1,202 @@
+ * hooks and assertions.
+ */
+public class MockExit extends Exit implements AutoCloseable {
+
+    @FunctionalInterface
+    public interface Procedure extends Exit.Procedure {
+        void execute(int statusCode, String message);
+        static Procedure noop() {
+            return (statusCode, message) -> { };
+        }
+    }
+
+    @FunctionalInterface
+    public interface ShutdownHookAdder extends Exit.ShutdownHookAdder {
+        void addShutdownHook(String name, Runnable runnable);
+
+        static ShutdownHookAdder noop() {
+            return (name, runnable) -> { };
+        }
+    }
```
Is the intent with these to eventually replace the superclass's `Procedure` and `ShutdownHookAdder` interfaces?

gharris1727 replied:
Yes, because these types are only useful for mocking out the exit behavior, and so should only be present in the `test` artifact.

C0urante replied:
Got it. Might be worth adding a comment to the superclass's interface stating that this is the intent, but up to you.

C0urante commented on clients/src/main/java/org/apache/kafka/common/utils/Exit.java:
```diff
@@ -119,4 +134,138 @@ public static void resetHaltProcedure() {
+            else
+                Runtime.getRuntime().addShutdownHook(new Thread(runnable));
+        }
+    }
```
These implementations match the `DEFAULT_HALT_PROCEDURE`,
`DEFAULT_EXIT_PROCEDURE`, and
`DEFAULT_SHUTDOWN_HOOK_ADDER` fields. Could we possibly de-duplicate them somehow?

C0urante commented on clients/src/main/java/org/apache/kafka/common/utils/Exit.java:
```diff
@@ -55,22 +55,37 @@ public interface ShutdownHookAdder {
     private static volatile Procedure haltProcedure = DEFAULT_HALT_PROCEDURE;
     private static volatile ShutdownHookAdder shutdownHookAdder = DEFAULT_SHUTDOWN_HOOK_ADDER;
 
+    /**
+     * Use of this method is discouraged, and should be replaced with {@link #exitOrThrow(int)}
+     */
```
For this method and the other newly-discouraged ones, would deprecation be an option? Or would that be noisier than it's worth?

gharris1727 replied:
Eventually yes. Once we remove most/all of these call-sites from AK code, we can deprecate them. I meant "discouraged" to mean "will be deprecated and removed eventually".

C0urante commented on clients/src/main/java/org/apache/kafka/common/utils/Exit.java:
```diff
@@ -119,4 +134,138 @@ public static void resetHaltProcedure() {
+     */
+    public abstract void addShutdownRunnable(String name, Runnable runnable);
+
+    private static final class SystemExit extends Exit {
```
Nit: it's a little strange to have `MockExit` in its own top-level file, but `SystemExit` as an inner class. I guess one reason for this could be to ensure that `SystemExit` is never accessed directly, but if we use the singleton pattern (ensuring at least that the class is never instantiated by external callers) and make it package-private, would that be sufficient to make sure it isn't abused? I'm also trying to see if we can reduce the size of the `Exit` class file to make it easier for new readers to grok.

C0urante commented on clients/src/test/java/org/apache/kafka/common/utils/MockExit.java:
```diff
@@ -0,0 +1,202 @@
+        for (Thread t : hooks) {
+            try {
+                t.join(timer.remainingMs());
+            } catch (InterruptedException ignored) {
```
Can/should we at least log a warning here?

C0urante commented on clients/src/test/java/org/apache/kafka/common/utils/MockExit.java:
```diff
@@ -0,0 +1,202 @@
+        try {
+            runAsserts();
+        } finally {
+            runHooks();
```
Could you possible give an overview of what kinds of shutdown hooks we add in our code base and whether we anticipate there might be some that it would not make sense to execute during tests?

gharris1727 replied:
It looks like most/all are intending to hook into ctrl-c used to shutdown a long-running process. This is how kafka, connect, mm1, mm2, streams, trogdor, console consumer, jmx reporter appear to work.

The MetadataShell deletes a temporary directory, and TestUtils temporary directories (e.g. kafka log directories) are deleted using the shutdown hook.

I think it looks harmless to run during tests, and if anything allows for resources to be cleaned up earlier than they are currently. Maybe the bigger risk is MockExit being used, and preventing these hooks from being executed later if the object is never close()d. Maybe ShutdownHookAdder.noop() is a bad default in that case, and calling the system implementation is a better default.

C0urante reviewed (commented):
Thanks Greg, I really like this and the care that was clearly taken to not only ease the migration but lead us to a better state at the end of it.

One higher-level thought I had: is it better to keep all of this new logic in the existing `Exit` class, or would it make sense to separate the two? IMO the final state we work towards should still be a single `Exit` class, but in the meantime, it could ease the migration and reduce complexity to separate the static (legacy) exit logic and the non-static (end state) logic. Not a blocker though, just a shower thought.

gharris1727 committed:
fixup: move SystemExit to toplevel, more javadocs

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 commented on clients/src/main/java/org/apache/kafka/common/utils/Exit.java:
```diff
@@ -119,4 +134,138 @@ public static void resetHaltProcedure() {
+     * {@link #setExitProcedure(Procedure)}, {@link #setHaltProcedure(Procedure)},
+     * or {@link #addShutdownHook(String, Runnable)} to avoid race conditions.
+     */
+    public static Exit staticContext() {
```
@C0urante I'll reproduce your comment on the other branch here to respond:

> Is Exit.staticContext() necessary here? I understand we might use it as a precautionary measure in case this constructor gets used in a testing context sometime in the future, but AFAICT it'd be acceptable to use Exit.system() right now since this doesn't appear to be used directly or indirectly in testing code.

You are correct, that the particular call-site for staticContext() that you commented on isn't ever called in a mocked-exit environment, so it would behave like calling system(). The commit I showed was close to the end of the migration, when it's possible that we could use system() in Connect.

I kept the staticContext() call because it's more defensive: if someone had called Exit.setExitProcedure, it would pick up the override and work correctly. Once we are sure there are no callers of setExitProcedure (e.g. no callers in Connect, no callers in Kafka, no callers in the whole ecosystem) we can swap staticContext() for system().

gharris1727 committed:
fixup: reorder methods to move modern, instance methods to the top of the class.

Signed-off-by: Greg Harris <greg.harris@aiven.io>

gharris1727 commented:
> One higher-level thought I had: is it better to keep all of this new logic in the existing Exit class, or would it make sense to separate the two? IMO the final state we work towards should still be a single Exit class, but in the meantime, it could ease the migration and reduce complexity to separate the static (legacy) exit logic and the non-static (end state) logic. Not a blocker though, just a shower thought.

I implemented this originally as "SafeExit"and didn't like it, and I mentioned why in the PR description. It was a silly name, and cause import churn that didn't feel worth it.

I do see where you're coming from, so I moved the instance methods up to the top of the class. This probably doesn't help javadoc users, but for source-code readers the instance methods should be more visible.

C0urante reviewed (commented):
LGTM, thanks Greg! Looking forward to seeing this new-and-improved API adapted across the code base.

gharris1727 committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-16420-add-safe-exit
